### PR TITLE
Split tests for setting session description into multiple files

### PR DESCRIPTION
**Pull request by soareschen**
The tests for setting session description is rather large, with 50+ steps involved in the spec. Currently the tests are located among RTCPeerConnection-setLocalDescription.html and RTCPeerConnection-setRemoteDescription.html, but the number of tests are getting large that it is becoming difficult to organize and keep track of the tests.

This PR is a succession of #6491 in effort to reorganize the session description tests. It split the tests further into smaller files. Specifically there are 4 more files for the combinations of setting local/remote offer/answer descriptions.

The tests are moved as is to the new files, with the coverage report file updated to reflect the changes.

### CODE REVIEW

File: webrtc/RTCPeerConnection-setLocalDescription-answer.html [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        "offer",
+        "pranswer",
+        "answer",
+        "rollback"
+      };
+   */
+
+  /*
+    4.3.1.6.  Set the RTCSessionSessionDescription
+      2.  If description is set as a local description, then run one of the following
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-setLocalDescription-answer.html]: `2.` => `2.2.2`

File: webrtc/RTCPeerConnection-setLocalDescription-answer.html [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+          assert_equals(pc.pendingLocalDescription, null);
+          assert_equals(pc.pendingRemoteDescription, null);
+        })));
+  }, 'setLocalDescription() with valid answer should succeed');
+
+  /*
+    4.3.2.  setLocalDescription
+      3.  Let lastAnswer be the result returned by the last call to createAnswer.
+      4.  If description.sdp is null and description.type is answer, set description.sdp
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-setLocalDescription-answer.html]: This completes all steps for `setLocalDescription`!!

File: webrtc/RTCPeerConnection-setLocalDescription-offer.html [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+   */
+
+  /*
+    4.3.2.  setLocalDescription
+      2.  Let lastOffer be the result returned by the last call to createOffer.
+      5.  If description.sdp is null and description.type is offer, set description.sdp
+          to lastOffer.
+
+    4.3.1.6.  Set the RTCSessionSessionDescription
+      2.2.  If description is set as a local description, then run one of the following
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-setLocalDescription-offer.html]: `2.2.` => `2.2.2.`

File: webrtc/coverage/set-session-description.txt [hunk 4]
```diff
@@ -236,7 +236,7 @@ https://w3c.github.io/webrtc-pc/archives/20170605/webrtc.html
 
 Coverage Report
 
-  Tested        28
-  Not Tested    22
+  Tested        30
+  Not Tested    20
   Untestable     8
   Total         58
```
Comment (rwaldron) [inline webrtc/coverage/set-session-description.txt]: Coverage spec updated to match

**Review by rwaldron — CHANGES_REQUESTED** [review]
Only minor spec numbering fixes and then this can land

File: webrtc/RTCPeerConnection-setLocalDescription-answer.html (hunk 2 above)
Comment (soareschen) [inline webrtc/RTCPeerConnection-setLocalDescription-answer.html]: Hope so. The tests have been so enormous that I have lost track of what's been covered countless times.

File: webrtc/RTCPeerConnection-setLocalDescription-answer.html (hunk 1 above)
Comment (soareschen) [inline webrtc/RTCPeerConnection-setLocalDescription-answer.html]: Fixed and rebased.

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/255509947)

# Firefox (nightly) #
Testing web-platform-tests at revision 44713b1fb271bbf881d7012725003132a39503c9
Using browser at version BuildID 20170719100147; SourceStamp 1b065ffd8a535a0ad4c39a912af18e948e6a42c1
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>6 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6537/webrtc/RTCPeerConnection-setLocalDescription-answer.html">/webrtc/RTCPeerConnection-setLocalDescription-answer.html</a></summary>


|                                                      Subtest                                                      | Results |                                                                                                                              Messages                                                                                                                             |
|-------------------------------------------------------------------------------------------------------------------|---------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                                                                   | OK      |                                                                                                                                                                                                                                                                   |
| `setLocalDescription() with valid answer should succeed`                                                          | FAIL    | `assert_not_equals: Expect session description to be defined, but got undefined got disallowed value undefined`                                                                                                                                                   |
| `setLocalDescription() with type answer and null sdp should use lastAnswer generated from createAnswer`           | FAIL    | `promise_test: Unhandled rejection with value: object "InvalidParameterError: Empty or null SDP provided to setLocalDescription"`                                                                                                                                 |
| `setLocalDescription() with answer not created by own createAnswer() should reject with InvalidModificationError` | FAIL    | `assert_throws: function "function() { throw e }" threw object "InvalidStateError: Calling SetLocal without first calling CreateOffer/Answer is not supported." that is not a DOMException InvalidModificationError: property "code" is equal to 11, expected 13` |
| `Calling setLocalDescription(answer) from stable state should reject with InvalidStateError`                      | FAIL    | `promise_test: Unhandled rejection with value: object "InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel."`                                                                                            |
| `Calling setLocalDescription(answer) from have-local-offer state should reject with InvalidStateError`            | FAIL    | `promise_test: Unhandled rejection with value: object "InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel."`                                                                                            |
| `Test onsignalingstatechange event for setLocalDescription() with valid answer should succeed`                    | PASS    |                                                                                                                                                                                                                                                                   |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6537/webrtc/RTCPeerConnection-setLocalDescription-offer.html">/webrtc/RTCPeerConnection-setLocalDescription-offer.html</a></summary>


|                                                     Subtest                                                     | Results |                                                                                                                              Messages                                                                                                                             |
|-----------------------------------------------------------------------------------------------------------------|---------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                                                                 | OK      |                                                                                                                                                                                                                                                                   |
| `setLocalDescription with valid offer should succeed`                                                           | FAIL    | `assert_not_equals: Expect session description to be defined, but got undefined got disallowed value undefined`                                                                                                                                                   |
| `setLocalDescription with type offer and null sdp should use lastOffer generated from createOffer`              | FAIL    | `promise_test: Unhandled rejection with value: object "InvalidParameterError: Empty or null SDP provided to setLocalDescription"`                                                                                                                                 |
| `setLocalDescription() with offer not created by own createOffer() should reject with InvalidModificationError` | FAIL    | `assert_throws: function "function() { throw e }" threw object "InvalidStateError: Calling SetLocal without first calling CreateOffer/Answer is not supported." that is not a DOMException InvalidModificationError: property "code" is equal to 11, expected 13` |
| `Set created offer other than last offer should reject with InvalidModificationError`                           | FAIL    | `assert_throws: function "function() { throw e }" threw object "InvalidSessionDescriptionError: Changing the media-type of m-sections is not allowed" that is not a DOMException InvalidModificationError: property "code" is equal to 0, expected 13`            |
| `Creating and setting offer multiple times should succeed`                                                      | FAIL    | `promise_test: Unhandled rejection with value: object "InvalidStateError: Cannot create offer in state have-local-offer"`                                                                                                                                         |
| `setLocalDescription(offer) should never resolve if connection is closed in parallel`                           | FAIL    | `assert_unreached: Pending promise should never be resolved. Instead it is rejected with: InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel. Reached unreachable code`                                 |
| `Test onsignalingstatechange event for setLocalDescription with valid offer should succeed`                     | PASS    |                                                                                                                                                                                                                                                                   |
| `Test onsignalingstatechange event for Creating and setting offer multiple times should succeed`                | PASS    |                                                                                                                                                                                                                                                                   |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6537/webrtc/RTCPeerConnection-setLocalDescription.html">/webrtc/RTCPeerConnection-setLocalDescription.html</a></summary>


|                                                                          Subtest                                                                          | Results |                                                     Messages                                                    |
|-----------------------------------------------------------------------------------------------------------------------------------------------------------|---------|-----------------------------------------------------------------------------------------------------------------|
|                                                                                                                                                           | OK      |                                                                                                                 |
| `Calling createOffer() and setLocalDescription() again after one round of local-offer/remote-answer should succeed`                                       | FAIL    | `assert_not_equals: Expect session description to be defined, but got undefined got disallowed value undefined` |
| `Switching role from answerer to offerer after going back to stable state should succeed`                                                                 | FAIL    | `assert_not_equals: Expect session description to be defined, but got undefined got disallowed value undefined` |
| `Test onsignalingstatechange event for Calling createOffer() and setLocalDescription() again after one round of local-offer/remote-answer should succeed` | PASS    |                                                                                                                 |
| `Test onsignalingstatechange event for Switching role from answerer to offerer after going back to stable state should succeed`                           | PASS    |                                                                                                                 |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6537/webrtc/RTCPeerConnection-setRemoteDescription-answer.html">/webrtc/RTCPeerConnection-setRemoteDescription-answer.html</a></summary>


|                                                  Subtest                                                  | Results |                                                                                Messages                                                                                |
|-----------------------------------------------------------------------------------------------------------|---------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                                                           | OK      |                                                                                                                                                                        |
| `setRemoteDescription() with valid state and answer should succeed`                                       | FAIL    | `assert_not_equals: Expect session description to be defined, but got undefined got disallowed value undefined`                                                        |
| `Calling setRemoteDescription(answer) from stable state should reject with InvalidStateError`             | FAIL    | `promise_test: Unhandled rejection with value: object "InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel."` |
| `Calling setRemoteDescription(answer) from have-remote-offer state should reject with InvalidStateError`  | FAIL    | `promise_test: Unhandled rejection with value: object "InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel."` |
| `Test onsignalingstatechange event for setRemoteDescription() with valid state and answer should succeed` | PASS    |                                                                                                                                                                        |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6537/webrtc/RTCPeerConnection-setRemoteDescription-offer.html">/webrtc/RTCPeerConnection-setRemoteDescription-offer.html</a></summary>


|                                                        Subtest                                                        | Results |                                                                                                              Messages                                                                                                             |
|-----------------------------------------------------------------------------------------------------------------------|---------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                                                                       | OK      |                                                                                                                                                                                                                                   |
| `setRemoteDescription with valid offer should succeed`                                                                | FAIL    | `assert_not_equals: Expect session description to be defined, but got undefined got disallowed value undefined`                                                                                                                   |
| `Setting remote description multiple times with different offer should succeed`                                       | FAIL    | `promise_test: Unhandled rejection with value: object "InvalidStateError: Cannot set remote offer or answer in current state have-remote-offer"`                                                                                  |
| `setRemoteDescription(offer) should never resolve if connection is closed in parallel`                                | FAIL    | `assert_unreached: Pending promise should never be resolved. Instead it is rejected with: InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel. Reached unreachable code` |
| `setRemoteDescription(offer) with invalid SDP should reject with RTCError`                                            | FAIL    | `assert_equals: Expect error detail field to set to sdp-syntax-error expected (string) "sdp-syntax-error" but got (undefined) undefined`                                                                                          |
| `Calling setRemoteDescription(offer) from have-local-offer state should reject with InvalidStateError`                | FAIL    | `promise_test: Unhandled rejection with value: object "InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel."`                                                            |
| `Test onsignalingstatechange event for setRemoteDescription with valid offer should succeed`                          | PASS    |                                                                                                                                                                                                                                   |
| `Test onsignalingstatechange event for Setting remote description multiple times with different offer should succeed` | PASS    |                                                                                                                                                                                                                                   |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6537/webrtc/RTCPeerConnection-setRemoteDescription.html">/webrtc/RTCPeerConnection-setRemoteDescription.html</a></summary>


|                                                                 Subtest                                                                  | Results |                                                                    Messages                                                                    |
|------------------------------------------------------------------------------------------------------------------------------------------|---------|------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                                                                                          | TIMEOUT |                                                                                                                                                |
| `setRemoteDescription with invalid type and invalid SDP should reject with TypeError`                                                    | PASS    |                                                                                                                                                |
| `setRemoteDescription() with invalid SDP and stable state should reject with InvalidStateError`                                          | PASS    |                                                                                                                                                |
| `Calling setRemoteDescription() again after one round of remote-offer/local-answer should succeed`                                       | FAIL    | `promise_test: Unhandled rejection with value: object "InvalidSessionDescriptionError: Remote description changes the media type of m-line 0"` |
| `Switching role from offerer to answerer after going back to stable state should succeed`                                                | FAIL    | `assert_not_equals: Expect session description to be defined, but got undefined got disallowed value undefined`                                |
| `Test onsignalingstatechange event for Calling setRemoteDescription() again after one round of remote-offer/local-answer should succeed` | TIMEOUT | `Test timed out`                                                                                                                               |
| `Test onsignalingstatechange event for Switching role from offerer to answerer after going back to stable state should succeed`          | PASS    |                                                                                                                                                |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/255509948)

# Sauce (safari) #
Testing web-platform-tests at revision 44713b1fb271bbf881d7012725003132a39503c9
Using browser at version 10.0
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>6 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6537/webrtc/RTCPeerConnection-setLocalDescription-answer.html">/webrtc/RTCPeerConnection-setLocalDescription-answer.html</a></summary>


|                                                      Subtest                                                      | Results |                 Messages                 |
|-------------------------------------------------------------------------------------------------------------------|---------|------------------------------------------|
|                                                                                                                   | OK      |                                          |
| `setLocalDescription() with valid answer should succeed`                                                          | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setLocalDescription() with type answer and null sdp should use lastAnswer generated from createAnswer`           | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setLocalDescription() with answer not created by own createAnswer() should reject with InvalidModificationError` | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Calling setLocalDescription(answer) from stable state should reject with InvalidStateError`                      | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Calling setLocalDescription(answer) from have-local-offer state should reject with InvalidStateError`            | FAIL    | `Can't find variable: RTCPeerConnection` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6537/webrtc/RTCPeerConnection-setLocalDescription-offer.html">/webrtc/RTCPeerConnection-setLocalDescription-offer.html</a></summary>


|                                                     Subtest                                                     | Results |                 Messages                 |
|-----------------------------------------------------------------------------------------------------------------|---------|------------------------------------------|
|                                                                                                                 | OK      |                                          |
| `setLocalDescription with valid offer should succeed`                                                           | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setLocalDescription with type offer and null sdp should use lastOffer generated from createOffer`              | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setLocalDescription() with offer not created by own createOffer() should reject with InvalidModificationError` | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Set created offer other than last offer should reject with InvalidModificationError`                           | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Creating and setting offer multiple times should succeed`                                                      | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setLocalDescription(offer) should never resolve if connection is closed in parallel`                           | FAIL    | `Can't find variable: RTCPeerConnection` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6537/webrtc/RTCPeerConnection-setLocalDescription.html">/webrtc/RTCPeerConnection-setLocalDescription.html</a></summary>


|                                                       Subtest                                                       | Results |                 Messages                 |
|---------------------------------------------------------------------------------------------------------------------|---------|------------------------------------------|
|                                                                                                                     | OK      |                                          |
| `Calling createOffer() and setLocalDescription() again after one round of local-offer/remote-answer should succeed` | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Switching role from answerer to offerer after going back to stable state should succeed`                           | FAIL    | `Can't find variable: RTCPeerConnection` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6537/webrtc/RTCPeerConnection-setRemoteDescription-answer.html">/webrtc/RTCPeerConnection-setRemoteDescription-answer.html</a></summary>


|                                                 Subtest                                                  | Results |                 Messages                 |
|----------------------------------------------------------------------------------------------------------|---------|------------------------------------------|
|                                                                                                          | OK      |                                          |
| `setRemoteDescription() with valid state and answer should succeed`                                      | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Calling setRemoteDescription(answer) from stable state should reject with InvalidStateError`            | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Calling setRemoteDescription(answer) from have-remote-offer state should reject with InvalidStateError` | FAIL    | `Can't find variable: RTCPeerConnection` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6537/webrtc/RTCPeerConnection-setRemoteDescription-offer.html">/webrtc/RTCPeerConnection-setRemoteDescription-offer.html</a></summary>


|                                                Subtest                                                 | Results |                 Messages                 |
|--------------------------------------------------------------------------------------------------------|---------|------------------------------------------|
|                                                                                                        | OK      |                                          |
| `setRemoteDescription with valid offer should succeed`                                                 | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Setting remote description multiple times with different offer should succeed`                        | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setRemoteDescription(offer) should never resolve if connection is closed in parallel`                 | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setRemoteDescription(offer) with invalid SDP should reject with RTCError`                             | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Calling setRemoteDescription(offer) from have-local-offer state should reject with InvalidStateError` | FAIL    | `Can't find variable: RTCPeerConnection` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6537/webrtc/RTCPeerConnection-setRemoteDescription.html">/webrtc/RTCPeerConnection-setRemoteDescription.html</a></summary>


|                                              Subtest                                               | Results |                 Messages                 |
|----------------------------------------------------------------------------------------------------|---------|------------------------------------------|
|                                                                                                    | OK      |                                          |
| `setRemoteDescription with invalid type and invalid SDP should reject with TypeError`              | FAIL    | `Can't find variable: RTCPeerConnection` |
| `setRemoteDescription() with invalid SDP and stable state should reject with InvalidStateError`    | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Calling setRemoteDescription() again after one round of remote-offer/local-answer should succeed` | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Switching role from offerer to answerer after going back to stable state should succeed`          | FAIL    | `Can't find variable: RTCPeerConnection` |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/255509954)

# Chrome (unstable) #
Testing web-platform-tests at revision 44713b1fb271bbf881d7012725003132a39503c9

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/255509955)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 44713b1fb271bbf881d7012725003132a39503c9
Using browser at version 14.14393
Starting 10 test iterations

File: webrtc/RTCPeerConnection-setLocalDescription-answer.html (hunk 2 above)
Comment (rwaldron) [inline webrtc/RTCPeerConnection-setLocalDescription-answer.html]: You're doing a great job

**Review by rwaldron — APPROVED** [review]
This needs to be rebased, but since you've done all the other work and I don't want to block this anymore I will do the rebase and get review from @bobholt

**Comment by rwaldron** [comment]
I rebased this for you and it's now merged! https://github.com/w3c/web-platform-tests/pull/6629

**Pull request closed without merging**